File: android/app/src/main/java/com/github/jachan/hallways/math/ContinuousStats.java
package com.github.jachan.hallways.math;

import static java.lang.Math.pow;
import static java.lang.Math.sqrt;

/**
 * Instead of recalculating the mean and standard devation based on all of the data, it is possible to continuously update those values.
 * Code to do so has been written here.
 * It follows http://www.jstor.org/stable/1266577
 * http://www.jstor.org.libproxy.utdallas.edu/stable/1266577
 */
public class ContinuousStats {
    public ContinuousStats() {
        M = 0.0d;
        S = 0.0d;
        ni = 0;
    }

    public void update(double xn) {
        ni++;
        double n = ni;
        S = S + (n - 1) / n  * pow(xn - M, 2.0d); // here M refers to m_(i - 1)
        M = (n - 1) / n * M + xn / n;
    }

    public double avg() {
        return M;
    }

    public double stddev() {
        double n = ni;
        return sqrt(S / n);
    }

    public int n() {
        return ni;
    }

    double M, S;
    int ni;
}

File: python_client/hallways/mystats/continuous_stats.py
import numpy as np

# It follows http://www.jstor.org/stable/1266577
# http://www.jstor.org.libproxy.utdallas.edu/stable/1266577
class ContinuousStats(object):
    def __init__(self):
        self._M = 0
        self._S = 0
        self._n = 0

    def update(self, xn):
        M, S, n = self._M, self._S, self._n

        n += 1
        S = S + (n - 1) / n * (xn - M)**2
        M = (n - 1) / n * M + xn / n

        self._M, self._S, self._n = M, S, n

    @property
    def avg(self):
        return self._M

    @property
    def stddev(self):
        if self._n == 0:
            return 0
        return np.sqrt(self._S / self._n)

    @property
    def n(self):
        return self._n

__all__ = ['ContinuousStats']
